bare_metal_tests/TESTS/bringup/uart/test.py

```diff
@@ -0,0 +1,330 @@
+    
+    shutil.copyfile(src, dst)
+    
+    logging.info('Wait %f sec after flashing.' % (delay))
+    time.sleep(delay)
+    
+def build_test(target, toolchain, clean_flag):
+    """Build the mbed test.
+    
+    @param target target id.
+    @param toolchain compile toolchain.
+    @return image path.
+    """
+    logging.info('Building mbed UART test for green tea.')
+    
+    clean = ''
+    if(clean_flag):
+        clean = '-c'
+    
+    cmd = 'mbed test -t %s -m %s --source . --source bare_metal_tests --build-data bare_metal_tests/TESTS/bringup/uart/build.log --compile %s' % (toolchain, target, clean)
+    return run_command(cmd)
+
+def open_port(port_id, baudrate):
+    """Open serial port.
+    
+    @param port_id serial port id.
+    @param baudrate serial port baudrate.
+    """
+    global serial_port
+    try:
+        serial_port = serial.Serial(
+        port=port_id,
+        baudrate=baudrate,
+        parity=serial.PARITY_NONE,
+        stopbits=serial.STOPBITS_ONE,
+        bytesize=serial.EIGHTBITS)
+
+    except:
+        logging.error('Could not open serial port.')
+        exit()
+
+def serial_write_line(line):
+    """Write line to the serial port.
+    
+    @param line line to be written.
+    """
+    serial_port.write(line)
+    logging.debug('[SEDN] %s' % line)
+
+def serial_read_line():
+    """Read line from the serial port with timeout.
+    
+    @return string read from the serial port.
+    """
+    global serial_port
+    limit = 0
+    line = str()
+    while(True):
+        if(serial_port.in_waiting == 0):
+            time.sleep(0.001)
+            limit += 1
+            if (limit == RETRY_LIMIT):
+                logging.debug('[RECV] %s' % line)
+                return line
+            continue
+        limit = 0
+        c = serial_port.read()
+        line += c
+        if(c == '\0'):
+            logging.debug('[RECV] %s' % line)
+            return line
+
+def wait_for_message_and_respond(msg):
+    """! This function waits for the message from the mbed device.
+         If valid message has been obtained, then it responds with
+         the same message. If invalid message has been received, then
+         it responds with 'RETRY\0'.
+         Note: First received message may be incomplete and is used for
+             synchronisation.
+             
+         @param msg expected message
+    """
+    global serial_port
+    for x in range(0, RETRY_LIMIT):
+        line = serial_read_line()
+        if(line == msg):
+            logging.info('Expected string has been received from the mbed device.')
+            serial_write_line(line)
+            return True
+        else:
+            # if the line is unexpected (corrupted) and nul terminator and send,
+            # this will trigger mbed part to resend the test message.
+            if not line.endswith('\0'):
+                line += '\0'
+            serial_write_line(line)
+        time.sleep(1)
```

Comment:
Is this sleep needed? your `serial_read_line` function already handles delays in serial reception.

Reply:
Sleep is not needed and has been removed.